### DISCUSSION TITLE
Scalability [graph usability] - Namespace Graphs

### DISCUSSION THREAD

**Discussion by jshaughn** [kiali/kiali#4508] [Roadmap]
This is related to the overall [Scalability Dessign Discussion](https://github.com/kiali/kiali/discussions/4080).

Graph Scalability is complicated, there are both usability and performance aspects.  The goal of this discussion is to see if we can improve usability for large topologies, by introducing a high-level `Namespace` graph type.  And then hopefully generate an issue to drive implementation.

For large topographies there are typically many namespaces, with substantial traffic between namespaces.  Note that the same namespace name on different clusters would likely be treated as different namespaces for this graph type.

The idea here would be a graph where each node represented a namespace and edges would reflect an aggregation of all inter-namespace traffic.  What this node would look like is something to be determined, but it could feasibly be a more rich node than our other node types, for example, it could be a group node that contained the set of workloads within the namespace, or some sort of other summary information about the namespace.  Health of the node would be applied at the namespace level, and edge-health would have to be for the new aggregation.

This could be a first step towards a more flexible graph that potentially allows expand/collapse of namespace nodes, perhaps even allowing each namespace to be applied with a different graph type (that’s a stretch).

This is scalability from a usability perspective; it allows users to interact with a less dense, high-level graph, and then to drill down to investigate areas of concern.

There are UX challenges here, but there are also challenges to make this a performant operation.  The assumption is that the graph would likely involve a  large number of namespaces, quite possibly all of the namespaces accessible to the user.  There is no really efficient way to query Prometheus for the relevant metrics, given the standard metric labels/dimensions provided by Istio.  What we really want is to gather cross-namespace telemetry, bounded by a set of namespaces.  Given a set of namespaces, S, there are only two ways to gather the required metrics, neither of which is efficient:

Fetch all metrics for NS in S via one query, using regex:

`namespace~= ns-1|ns-2|..|ns-n`

Fetch all metrics for N in S via one query, using OR:

`namespace=ns-1 OR namespace=ns-2 OR … namespace=ns-n`

Not only is the fetch inefficient, it pulls all of the metrics, and then in the fetching code we’d need to throw away the intra-namespace traffic.  (To be honest, it’s worse than this, because the above example is simplified, we’d actually need to consider the source_namespaces and destination_namespaces.

What we really want is to be able to fetch metrics where source_namespace != destination_namespace.  Note that Prometheus does not support queries that compare labels within the same time_series, as it goes against the design principles, and would be inefficient even if it were allowed.

To support namespace graphs I think we’d need to augment the default telemetry in Istio, and add a new dimension, something like:

`cross_namespace: source_ns != dest_ns || source_cluster != dest_cluster`

This is a departure to our history of using default metrics.  We could ask Istio to add it as a default dimension, but otherwise we would have to get users to apply the change in order to unlock the new graph type.  Some good news is that it would not increase time-series cardinality.

This new dimension would provide a much more efficient query, something like:

`cross_namespace=true`

That should be able to make the proposed graph type very feasible.

Thoughts?

**Comment by leandroberetta** [comment]
It would be very nice to have this feature. 

Regarding augmenting the metrics, sounds that is necessary but my only concern is that it will make updates more difficult (any updates to Prometheus Kiali's metrics).

To try this kind of scenarios, multiples namespaces with many services, I'm working on a demo where you will be able to say "I want 10 namespaces, with 10 services on each" and it will generate random connections between inter namespace services.

**Reply by jshaughn** [reply]
That demo will be very useful, I think.

**Comment by nrfox** [comment]
I like this feature and I think it would be great to narrow the scope of the graph to more visually consumable defaults. Perhaps namespace boxes can be collapsed by default if they contain more than X number of nodes? Figuring out what's the max number of nodes that can be displayed before the graph becomes visually unusable is probably a separate issue. I also like the idea of stripping out visual information from the graph when you are zoomed out and adding more progressively as you zoom in like what @lucasponce has mentioned in the past. Maybe this feature can work in that same way.

For the metrics side of things, does Kiali need to query anything differently for this or can these just be aggregated client side by adding up the node/edges for all elements in the namespace? Is the way that Kiali queries metrics today for separate namespaces a bottleneck?

**Reply by jshaughn** [reply]
>  Perhaps namespace boxes can be collapsed by default if they contain more than X number of nodes?

Yes, this has been something on my mind as well.  It's why I was playing with that expand/collapse stuff a couple of weeks ago.

>  I also like the idea of stripping out visual information from the graph when you are zoomed out 

Yes, when labeling/badging gets too small it's not useful and just makes things more busy.

> For the metrics side of things, does Kiali need to query anything differently for this or can these just be aggregated client side

I think for this to be scalable from a perf perspective, we do need to have dedicated back-end code and queries because generating namespace graphs for a lot of namespaces is time-consuming.

**Comment by jmazzitelli** [comment]
Why not just think about a "namespace" graph that simply has namespace nodes with aggregated edges? I wouldn't worry about collapsing them or adding workload nodes inside the namespace node. Just have a graph with namespace nodes inter-connected with each other and the side panel showing data for the selected node (just like our service-only graph type).

**Reply by jshaughn** [reply]
> Why not just think about a "namespace" graph that simply has namespace nodes with aggregated edges? 

Yes, that's exactly what I'm proposing here.  A next step would be to figure out how to best drill into a namespace, or namespaces of interest.  The challenge is what I tried to describe, to do that aggregation is not performant with the standard metrics.  At least not in any way I can see.

**Comment by lucasponce** [comment]
To introduce a new view/graph focused on namespaces as nodes is a nice idea, but I think that users will like to see a combined scenario, where you may want to collapse a namespace just seeing a box meanwhile in other namespaces you are interested to see the details.

Note that this is a very common pattern in other kubernetes topologies tooling (I remember nice examples in tooling from ACS or Cilium).

I think it's a point where we can build a solid model detached from the implementation that may help us to differentiate two aspects:
- How the model is built and cached.
- How the model is queried and manipulated.

With this approach we could:
- Keep the current model to build the topology graph, but introduce significant changes there in a transparent way (more caching, use other sources of information to build the dependencies and relations between nodes).
- Let the user to operate the graph/topology at scale, allowing to show namespace, expand/collapse, filtering, zooming, in a very fast way, operating with a cached model (configurable by the user).

In short, I'm not opposed to introduce a new namespace graph a short step, but I think this proposal also revisit the discussion we started in June about the same topic.

Not sure if my comment is more orthogonal to the namespace graph discussion and it may deserve another discussion, but I see it related.

**Comment by lucasponce** [comment]
> There are UX challenges here, but there are also challenges to make this a performant operation. The assumption is that the graph would likely involve a large number of namespaces, quite possibly all of the namespaces accessible to the user. There is no really efficient way to query Prometheus for the relevant metrics, given the standard metric labels/dimensions provided by Istio. What we really want is to gather cross-namespace telemetry, bounded by a set of namespaces. Given a set of namespaces, S, there are only two ways to gather the required metrics, neither of which is efficient:

I think this is one of the areas that we need to discuss. 
On large topologies the way to generate a graph should change and potentially it should provide another type of graph that is focused in dependencies rather than traffic.

I think the current graph is perfect when the user wants to position in a single element, to see all incoming/outgoing connections, but as soon as it grows it's pretty inneficient, perhaps more in the UI side https://github.com/kiali/kiali/discussions/4518 rather than the backen pressure.

So, I think we'd need to introduce a new model that allows to:

- Use the workloads as main source/backbone for this new graph.
- One the main backbone of the model is defined, a better/intelligent layout can be defined as discussed https://github.com/kiali/kiali/discussions/4518
- The health of the elements should be first information that user would like to see.
- Connections between elements can be on/off and being enabled on zoom-in, or selecting a particular area.
- Using the main workloads as base for the backbone would help to add more value about what to show (i.e. filter per user labelling, show me all of these workloads, then, show me their connections).

P.S.: I have impression that this issue and #4518 are pretty close/similar, so probably we may talk about both topics.
